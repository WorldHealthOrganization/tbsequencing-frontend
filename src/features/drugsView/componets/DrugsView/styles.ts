import { css } from '@emotion/react';

const wrapper = css({
  padding: '40px 54px',
});

const header = css({
  marginBottom: '24px',
});

export {
  wrapper, header,
};
