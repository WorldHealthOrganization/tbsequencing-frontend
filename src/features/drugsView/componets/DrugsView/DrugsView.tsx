/** @jsxImportSource @emotion/react */
import React, { useEffect, useMemo, useState } from 'react';
import { useSearchParams } from 'react-router-dom';
import H1 from '../../../../components/typography/H1';
import { header, wrapper } from './styles';
import ButtonsGroup from '../ButtonsGroup/ButtonsGroup';
import DrugDescriptionWrapper from '../DrugDescriptionWrapper/DrugDescriptionWrapper';
import { ChartDataType, Regions, ResistantType } from '../../models';
import {
  useGetChartDataQuery,
  useGetDrugsDataQuery,
  useGetTableDataQuery,
  useGetGeneAssociationsGroupedQuery,
} from '../../../../services/drugsApi/drugsApi';
import { IChartDataLocal, IDrug } from '../../../../services/drugsApi/models';
import { DEFAULT_PAGE_SIZE, useTableData } from '../../hooks/useTableData';
import DataGrid from '../../../../components/DataGrid';
import { DrugTableColumns, GridType, IColumn } from '../../../../components/DataGrid/models';
import { pagination } from '../../../../components/DataGrid/styles';
import Pagination from '../Pagination';
import {
  createQueryParamsFromRegion,
} from '../../../../services/drugsApi/utils';
import ChartWrapper from '../Chart/ChartWrapper';

export const columns: IColumn<DrugTableColumns>[] = [
  { name: 'geneName', header: 'Gene' },
  { name: 'variantName', header: 'Genomic coordinates' },
  { name: 'nucleodicAnnName', header: 'Nucleic Annotation' },
  { name: 'proteicAnnName', header: 'Proteic Annotation' },
  { name: 'consequence', header: 'Consequence' },
  { name: 'variantGrade', header: 'Grade' },
  { name: 'globalFrequency', header: 'Global Frequency' },
  { name: 'totalCounts', header: 'Total Samples' },
  { name: 'resistantCount', header: 'Resistant Count' },
  { name: 'susceptbleCount', header: 'Susceptible Count' },
  { name: 'intermediateCount', header: 'Intermediate Count' },
];

const DrugsView = () => {
  const [searchParams] = useSearchParams();

  const defaultDrug = {
    code: searchParams.get('code'),
    drugId: Number(searchParams.get('drugId')),
    drugName: searchParams.get('drugName'),
  };

  const [selectedDrug, setSelectedDrug] = useState<IDrug | {}>(
    defaultDrug.drugId ? defaultDrug : {},
  );

  const [resistantType, setResistantType] = useState<ResistantType>(ResistantType.Phenotypic);
  const [selectedDates] = useState<number[]>([1964, new Date().getFullYear()]);
  const [selectedRegion, setSelectedRegion] = useState(Regions.ALL);
  const [dataType, setDataType] = useState<ChartDataType>(ChartDataType.Ratio);

  const {
    data: drugsData = [],
  } = useGetDrugsDataQuery({ isAssociated: 1 });

  const {
    data: chartData = [],
    isLoading: isChartDataLoading,
  } = useGetChartDataQuery({
    resistantType,
    year: selectedDates,
    countryId: createQueryParamsFromRegion(selectedRegion),
  });

  const {
    data: genesAssociationsData,
    isFetching: isAssociationLoading,
  } = useGetGeneAssociationsGroupedQuery(
    { drugId: (selectedDrug as IDrug).drugId },
    { skip: !(selectedDrug as IDrug).drugId },
  );

  const {
    rowsCount,
    tableData = [],
    isTableDataLoading,
    sortState,
    filters,
    page,
    updateSort,
    updateFilters,
    updatePaging,
  } = useTableData({
    drugID: (selectedDrug as IDrug).drugId,
    useGetDataQuery: useGetTableDataQuery,
  });

  useEffect(() => {
    if (drugsData.length && !(selectedDrug as IDrug).drugId) {
      setSelectedDrug(drugsData[0]);
    }
  }, [drugsData]);

  const formattedChartData = useMemo(() => {
    if (!chartData) return [];
    return chartData.reduce((acc: IChartDataLocal[], chartItem) => {
      if (drugsData.length) {
        const currentDrug = drugsData.find((drug: IDrug) => drug.drugId === chartItem.drug);
        if (currentDrug) {
          acc.push({
            ...chartItem,
            drug: currentDrug,
            ratioSusceptible: parseFloat(chartItem.ratioSusceptible).toFixed(1),
            ratioResistant: parseFloat(chartItem.ratioResistant).toFixed(1),
            ratioIntermediate: parseFloat(chartItem.ratioIntermediate).toFixed(1),
            drugCode: currentDrug.code,
          });
        }
      }
      return acc;
    }, []);
  }, [chartData, drugsData]);

  const handleResistantTypeChange = (type: ResistantType) => {
    setResistantType(type);
  };

  // const handleTimePeriodChange = (value: number[]) => {
  //   setSelectedDates(value);
  // };

  const handleRegionChange = (value: Regions) => {
    setSelectedRegion(value);
  };

  const handleDataTypeChanged = (value: ChartDataType) => {
    setDataType(value);
  };

  const handleSelectDrugChange = (value: number) => {
    const selectedDrugData = drugsData.find((drug) => drug.drugId === value) as IDrug;
    setSelectedDrug(selectedDrugData);
  };

  return (
    <div css={wrapper}>
      <H1 style={header}>Resistance overview per drug</H1>
      <ButtonsGroup
        resistantType={resistantType}
        chartDataType={dataType}
        resistantTypeClick={handleResistantTypeChange}
        // setSelectedDates={handleTimePeriodChange}
        handleRegionChange={handleRegionChange}
        handleDataTypeChanged={handleDataTypeChanged}
        selectedRegion={selectedRegion}
        // selectedDates={selectedDates}
      />
      <ChartWrapper
        isLoading={isChartDataLoading}
        dataType={dataType}
        data={formattedChartData}
        selectedDrug={selectedDrug as IDrug}
        onClick={(data) => setSelectedDrug(data.drug || {})}
      />
      <DrugDescriptionWrapper
        handleSelectDrugChange={handleSelectDrugChange}
        drugs={drugsData}
        associations={genesAssociationsData}
        isLoadingAssociations={isAssociationLoading}
        selectedDrug={selectedDrug as IDrug}
      >
        <DataGrid<DrugTableColumns>
          isLoading={isTableDataLoading}
          columns={columns}
          dataSource={tableData}
          filterValues={filters}
          onFilterChange={updateFilters}
          onSortingChange={updateSort}
          selectedDropdownItem={selectedDrug}
          sortState={sortState}
          type={GridType.Drug}
        />
        <div css={pagination.wrapper}>
          <Pagination
            page={page}
            pageCount={Math.ceil(rowsCount / DEFAULT_PAGE_SIZE)}
            onPageClick={updatePaging}
          />
        </div>
      </DrugDescriptionWrapper>
    </div>
  );
};

export default DrugsView;
