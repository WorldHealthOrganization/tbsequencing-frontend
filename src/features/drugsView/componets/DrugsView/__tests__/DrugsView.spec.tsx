import React from 'react';
import {
  getTreeJSONWithRedux, render, screen, wrappedInRouter,
} from '../../../../../utils/testUtils/jestUtils';
import DrugsView from '../DrugsView';

describe('DrugsView', () => {
  test('renders DrugsView components without error', () => {
    render(
      wrappedInRouter(<DrugsView />),
    );
  });

  test('renders DrugsView title without error', () => {
    render(
      wrappedInRouter(<DrugsView />),
    );

    screen.getByText('Resistance overview per drug');
  });

  test('Should render correctly', () => {
    const tree = getTreeJSONWithRedux(
      wrappedInRouter(<DrugsView />),
    );

    expect(tree).toMatchSnapshot();
  });
});
