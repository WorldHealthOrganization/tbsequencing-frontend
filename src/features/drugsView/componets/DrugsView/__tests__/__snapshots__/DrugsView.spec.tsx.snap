// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrugsView Should render correctly 1`] = `
<div
  className="css-11lacbf-DrugsView"
>
  <p
    className="MuiTypography-root MuiTypography-body1 css-euu0lb-MuiTypography-root-BaseText"
  >
    Resistance overview per drug
  </p>
  <div
    className="css-1667gqu-ButtonsGroup"
  >
    <div
      className="css-1pim0p0-ButtonsGroup"
    >
      <button
        className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButtonBase-root css-1ve9vu2-MuiButtonBase-root-MuiButton-root-AppButton"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-3cwtjx-MuiTypography-root-BaseText"
        >
          Phenotypic resistance
        </p>
      </button>
      <button
        className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-disableElevation MuiButtonBase-root css-sfprtt-MuiButtonBase-root-MuiButton-root-AppButton"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-3cwtjx-MuiTypography-root-BaseText"
        >
          Genotypic resistance
        </p>
      </button>
    </div>
    <div
      className="css-1pim0p0-ButtonsGroup"
    >
      <button
        className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-disableElevation MuiButtonBase-root css-sfprtt-MuiButtonBase-root-MuiButton-root-AppButton"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-3cwtjx-MuiTypography-root-BaseText"
        >
          WHO region
        </p>
        <span
          className="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
        >
          <span
            aria-hidden={true}
            className="material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium css-kp9ftd-MuiIcon-root"
          >
            arrow_drop_down
          </span>
        </span>
      </button>
    </div>
    <div
      className="css-1pim0p0-ButtonsGroup"
    >
      <button
        className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButtonBase-root css-1ve9vu2-MuiButtonBase-root-MuiButton-root-AppButton"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-3cwtjx-MuiTypography-root-BaseText"
        >
          Proportions
        </p>
      </button>
      <button
        className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-disableElevation MuiButtonBase-root css-sfprtt-MuiButtonBase-root-MuiButton-root-AppButton"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-3cwtjx-MuiTypography-root-BaseText"
        >
          Absolute numbers
        </p>
      </button>
    </div>
  </div>
  <div
    className="css-1ks2rbf-ChartWrapper"
  >
    <span
      className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1ag5kwf-MuiCircularProgress-root"
      role="progressbar"
      style={
        Object {
          "height": 40,
          "width": 40,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </span>
  </div>
  <div />
</div>
`;
